File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.py
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        def backtrack(r, c, total):
            nonlocal result
            result = max(result, total)
            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]
                if (nr < 0 or len(grid) <= nr or
                    nc < 0 or len(grid[0]) <= nc or
                    (nr, nc) in chk or
                    grid[nr][nc] == 0): continue
                chk.add((nr, nc))
                backtrack(nr, nc, total + grid[nr][nc])
                chk.discard((nr, nc))


        dr = [0, 0, -1, 1]
        dc = [1, -1, 0, 0]

        result = 0
        chk = set()

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] != 0:
                    chk.add((i, j))
                    backtrack(i, j, grid[i][j])
                    chk.discard((i, j))

        return result

File: 1219-path-with-maximum-gold/1219-path-with-maximum-gold.java
class Solution {
    private int result;
    private boolean[][] chk;
    private int[] dr = {0, 0, -1, 1};
    private int[] dc = {1, -1, 0, 0};

    private void backtrack(int r, int c, int total, int[][] grid) {
        result = Math.max(result, total);
        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i], nc = c + dc[i];
            if (nr < 0 || grid.length <= nr ||
                nc < 0 || grid[0].length <= nc ||
                chk[nr][nc] || grid[nr][nc] == 0) continue;
            chk[nr][nc] = true;
            backtrack(nr, nc, total + grid[nr][nc], grid);
            chk[nr][nc] = false;
        }
    }

    public int getMaximumGold(int[][] grid) {
        result = 0;
        chk = new boolean[grid.length][grid[0].length];

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] != 0) {
                    chk[i][j] = true;
                    backtrack(i, j, grid[i][j], grid);
                    chk[i][j] = false;
                }
            }
        }

        return result;
    }
}